### ISSUE TITLE
Use flycheck for display of compiler notes

### ISSUE THREAD

**Issue by lunaryorn** [ensime/ensime-server#932]
Hello,

is there a way to disable the compiler notes highlighting in Ensime, i.e. tell Ensime that it shouldn't highlight compiler errors in the buffer?

Background:  I maintain [Flycheck](http://www.flycheck.org), an Emacs extension for background syntax checking, which pretty much does the same, but with more features—notably, a error list.  I'd like to use Flycheck to highlight Ensime errors, but that only makes sense when Ensime doesn't highlight errors by itself.

**Comment by fommil** [comment]
I think this ticket should be renamed to "use flycheck to display the compiler notes" if you think that would be possible? :smile: There is no point in us having a separate system.

The way ENSIME works is that the server asynchronously sends the notes in SWANK messages, so in emacs we have s-expressions containing the notes for each buffer. Is there a way to send that information to flycheck for display?

**Comment by fommil** [comment]
btw, this would dovetail with #384 which would be another source of compiler warnings/errors.

**Comment by lunaryorn** [comment]
@fommil Being the maintainer of Flycheck I know that it is possible, but I'd not like to have Ensime depend on Flycheck.  Flycheck—and specifically that part of Flycheck I'd use for this purpose—isn't stable yet.

I'd would prefer if Flycheck integration was optional, and a separate package on MELPA.  And I'd prefer to maintain it myself, outside of Ensime, for I'd like to have the freedom to change Flycheck's API without prior notice and without taking care for any larger project if I find that the API needs improvement.

For this reason, I don't want to replace Ensime's built-in reporting, and I would not recommend you to do so as well.  I'd really just like to have a way to disable Ensime's built-in reporting, so that it doesn't interfere with an Ensime addon for Flycheck.

**Comment by fommil** [comment]
that's a shame. In that case you'd probably be best to submit a PR which has a new user variable to disable the rendering of the notes.

If it helps, we'd be quite happy for you to take over the ownership of a flycheck file within the ENSIME repo and give you merging rights.

**Comment by lunaryorn** [comment]
I'd prefer to keep the library in a separate repo until it is ready for a first release, and then release as a separate package initially.  If it turns out to work well, and if there are no issues, we can still merge it into Ensime proper, and turn the old package into a simple dummy package with a dependency on Ensime.

But I don't want to break anything in Ensime.

**Comment by fommil** [comment]
ok, if that works for you. FYI we'd be happy to have the code within ensime, activated with an experimental flag.

**Comment by lunaryorn** [comment]
@fommil Well, then, if you're open to unstable code, I'll add it to Ensime.  But let me first come up with a working solution :)  I only have a rough draft yet, completely untested, and likely broken.  Give me a few days to bring it into a somewhat decent shape.  Probably till the weekend, it's going to be a busy week for me.

**Comment by fommil** [comment]
cool! Adding a test to the test suite shouldn't be too onerous for you, and so long as it needs a user flag to enable it, nobody will be affected.

**Comment by lunaryorn** [comment]
@fommil Sorry for the delay, but I have a status update:  I've got a first working version of the a Flycheck checker for Ensime ready at https://github.com/lunaryorn/.emacs.d/blob/master/lisp/flycheck-ensime.el—in theory.

Practically it doesn't really work, though  It parses and highlights Ensime's compiler notes, but frequently causes Emacs to hang.  I don't know why, yet, but I'll try to debug this issue and fix it.

I also need to change Flycheck to work better with external backends that do their own syntax checking like Ensime.  With these backends, Flycheck should try to take control of syntax checking, and instead just wait passively for incoming errors from the external backend.  Unfortunately, that's not really supported currently.

TL;DR: I've made progress, but it'll take some time until it's ready for production.

**Comment by fommil** [comment]
:smile:

**Comment by fommil** [comment]
This issue was moved to ensime/ensime-emacs#160
